internal/controllers/machineset/machineset_controller_status.go

```diff
@@ -0,0 +1,314 @@
+	}
+	gotReplicas := ms.Status.Replicas
+
+	referencesMessage := calculateMissingReferencesMessage(ms, bootstrapObjectNotFound, infrastructureObjectNotFound)
+
+	// Not scaling up or in deletion.
+	if gotReplicas >= desiredReplicas {
+		var message string
+		if referencesMessage != "" {
+			message = fmt.Sprintf("ScalingUp can't happen %s", referencesMessage)
+		}
+		v1beta2conditions.Set(ms, metav1.Condition{
+			Type:    clusterv1.MachineSetScalingUpV1Beta2Condition,
+			Status:  metav1.ConditionFalse,
+			Reason:  clusterv1.MachineSetNotScalingUpV1Beta2Reason,
+			Message: message,
+		})
+		return
+	}
+
+	message := fmt.Sprintf("ScalingUp from %d to %d replicas", gotReplicas, desiredReplicas)
+	if referencesMessage != "" {
+		message = fmt.Sprintf("%s is blocked %s", message, referencesMessage)
+	}
+	// Scaling up.
+	v1beta2conditions.Set(ms, metav1.Condition{
+		Type:    clusterv1.MachineSetScalingUpV1Beta2Condition,
+		Status:  metav1.ConditionTrue,
+		Reason:  clusterv1.MachineSetScalingUpV1Beta2Reason,
+		Message: message,
+	})
+}
+
+func setScalingDownCondition(ms *clusterv1.MachineSet, machines []*clusterv1.Machine, bootstrapObjectNotFound, infrastructureObjectNotFound bool, owner *clusterv1.MachineDeployment) {
+	// If we got unexpected errors in listing the machines (this should happen rarely), surface them
+	if machines == nil {
+		v1beta2conditions.Set(ms, metav1.Condition{
+			Type:    clusterv1.MachineSetScalingDownV1Beta2Condition,
+			Status:  metav1.ConditionUnknown,
+			Reason:  clusterv1.MachineSetScalingDownInternalErrorV1Beta2Reason,
+			Message: "Please check controller logs for errors",
+		})
+		return
+	}
+
+	// Surface if .spec.replicas is not yet set (this should happen rarely).
+	// This could e.g. be the case when using autoscaling with MachineSets.
+	if ms.Spec.Replicas == nil {
+		v1beta2conditions.Set(ms, metav1.Condition{
+			Type:   clusterv1.MachineSetScalingDownV1Beta2Condition,
+			Status: metav1.ConditionUnknown,
+			Reason: clusterv1.MachineSetScalingDownWaitingForReplicasToBeSetV1Beta2Reason,
+		})
+		return
+	}
+
+	desiredReplicas := *ms.Spec.Replicas
+	// Deletion is equal to 0 desired replicas.
+	if !ms.DeletionTimestamp.IsZero() {
+		desiredReplicas = 0
+	}
+
+	// Scaling down.
+	if int32(len(machines)) > (desiredReplicas) {
+		messages := []string{fmt.Sprintf("ScalingDown from %d to %d replicas", len(machines), desiredReplicas)}
+
+		machinesStaleDeleting := []string{}
+
+		for _, machine := range machines {
+			if !machine.GetDeletionTimestamp().IsZero() && time.Since(machine.GetDeletionTimestamp().Time) >= time.Minute*30 {
+				machinesStaleDeleting = append(machinesStaleDeleting, machine.GetName())
+			}
+		}
+
+		if len(machinesStaleDeleting) > 0 {
+			messages = append(messages, fmt.Sprintf("Machines stuck in deletion for more than 30 minutes: %s", clog.StringListToString(machinesStaleDeleting)))
+		}
+
+		// Scaling is blocked when the bootstrap or infrastructure objects do not exist.
```

Comment:
I think scale down is not blocked. If I see correctly this PR changes that (which is a good thing)

(If I'm correct, I would suggest to not include this message in scale down. If it doesn't matter for scale down, let's not confuse users)